### PR TITLE
executor,infoschema: Add pad_attribute to collation info | tidb-test=pr/2354

### LINKED ISSUE #54250: SHOW COLLATION format does not compatible with mysql 8.0
### 1. Minimal reproduce step (Required)
```
mysql> SHOW COLLATION WHERE `Charset` = 'utf8mb4' AND `Default` = 'Yes';
+-------------+---------+------+---------+----------+---------+
| Collation   | Charset | Id   | Default | Compiled | Sortlen |
+-------------+---------+------+---------+----------+---------+
| utf8mb4_bin | utf8mb4 |   46 | Yes     | Yes      |       1 |
+-------------+---------+------+---------+----------+---------+
1 row in set (0.00 sec)

mysql> select @@version;
+--------------------+
| @@version          |
+--------------------+
| 8.0.11-TiDB-v8.1.0 |
+--------------------+
1 row in set (0.01 sec)

```

### 2. What did you expect to see? (Required)
The output format should be same as mysql8.0, which is:
```
mysql> SHOW COLLATION WHERE `Charset` = 'utf8mb4' AND `Default` = 'Yes';
+--------------------+---------+-----+---------+----------+---------+---------------+
| Collation          | Charset | Id  | Default | Compiled | Sortlen | Pad_attribute |
+--------------------+---------+-----+---------+----------+---------+---------------+
| utf8mb4_0900_ai_ci | utf8mb4 | 255 | Yes     | Yes      |       0 | NO PAD        |
+--------------------+---------+-----+---------+----------+---------+---------------+
1 row in set (0.01 sec)

mysql> select @@version;
+-----------+
| @@version |
+-----------+
| 8.0.37    |
+-----------+
1 row in set (0.00 sec)
```
### 3. What did you see instead (Required)
mysql `show collation` command returns 7 columns. But tidb `show collation` command returns 6 columns.

### 4. What is your TiDB version? (Required)
```
select @@version;
+--------------------+
| @@version          |
+--------------------+
| 8.0.11-TiDB-v8.1.0 |
+--------------------+
1 row in set (0.01 sec)
```
Which claims tidb is compatible with mysql 8.0.11

### PR DESCRIPTION
**Pull request by dveeden**
### What problem does this PR solve?

Issue Number: close #54250 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
A `pad_attribute` column has been included in the `information_schema.collations` table and in the output of `SHOW COLLATION` to improve MySQL compatibility.
```

### CODE REVIEW

**Comment by kennytm** [comment]
`Sortlen` in MySQL is defined as "*related to the amount of memory required to sort strings expressed in the character set*", but the meaning of the value is never explained. According to https://dba.stackexchange.com/questions/281360/whats-sortlen-of-mysql-collation

> After 20 years of using MySQL, I have found no end-user use for `Sortlen`. Ignore it.

If we need to copy the result exactly from MySQL then many of the filled in sortlen are wrongly set to 1 in this PR:

```
mysql> select any_value(collation_name), sortlen, count(*) from information_schema.collations group by sortlen;
+---------------------------+---------+----------+
| any_value(collation_name) | sortlen | count(*) |
+---------------------------+---------+----------+
| armscii8_general_ci       |       1 |       98 |
| cp1250_czech_cs           |       2 |        3 |
| gb18030_unicode_520_ci    |       8 |      121 |
| latin2_czech_cs           |       4 |        2 |
| utf8mb4_0900_ai_ci        |       0 |       61 |
| utf8mb4_ja_0900_as_cs_ks  |      24 |        1 |
+---------------------------+---------+----------+
6 rows in set (0.02 sec)
```

OTOH if we accept `sortlen` is meaningless then keeping them as `1` should be good enough.

**Comment by dveeden** [comment]
I set the sortlen for the collations in the new framework only.

**Review by tiancaiamao — APPROVED** [review]

**Comment by D3Hunter** [comment]
@hawkingrei @lcwangchao

**Review by hawkingrei — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

File: pkg/planner/core/planbuilder.go [hunk 1]
```diff
@@ -5160,9 +5162,10 @@ func buildShowSchema(s *ast.ShowStmt, isView bool, isSequence bool) (schema *exp
 	case ast.ShowVariables, ast.ShowStatus:
 		names = []string{"Variable_name", "Value"}
 	case ast.ShowCollation:
-		names = []string{"Collation", "Charset", "Id", "Default", "Compiled", "Sortlen"}
+		names = []string{"Collation", "Charset", "Id", "Default", "Compiled", "Sortlen", "Pad_attribute"}
 		ftypes = []byte{mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeLonglong,
-			mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeLonglong}
+			mysql.TypeVarchar, mysql.TypeVarchar, mysql.TypeLonglong, mysql.TypeVarchar}
+		flags = []uint{0, 0, mysql.UnsignedFlag | mysql.NotNullFlag, 0, 0, 0, 0}
```
Comment (Defined2014) [inline pkg/planner/core/planbuilder.go:5168]: Seems `UnsignedFlag` will cause JDBC-5.x startup failed.

Comment (dveeden) [reply pkg/planner/core/planbuilder.go:5168]: @Defined2014 Do you have more info like versions and a stacktrace? Maybe a testcase I can run? Might be good to create a new issue for this?

Comment (dveeden) [reply pkg/planner/core/planbuilder.go:5168]: And as JDBC 4.3 is the latest JDBC version, I assume you meant MySQL Connector/J 5.x ?

Comment (Defined2014) [reply pkg/planner/core/planbuilder.go:5168]: Yes, MySQL Connector/J 5.x. I only have a Chinese document, maybe you can use a translator to read it. https://www.cnblogs.com/youzhibing/p/18104262

Comment (dveeden) [reply pkg/planner/core/planbuilder.go:5168]: The translation feature in Firefox seems to do a good job on that document. Thanks.

Comment (kennytm) [reply pkg/planner/core/planbuilder.go:5168]: off-topic but it's interesting they have to dig deep into the Java code to learn something that is [clearly documented](https://dev.mysql.com/doc/connectors/en/connector-j-reference-type-conversions.html) :eyes:

| MySQL Type Name | Return value of `GetColumnTypeName` | Return value of `GetColumnClassName` |
|---|---|---|
| `BIGINT[(M)]` | `BIGINT` | `java.lang.Long` |
| `BIGINT[(M)] UNSIGNED` | `BIGINT UNSIGNED` | `java.math.BigInteger` |

**Pull request merged**